lib/Analysis/Utility.cpp

```diff
@@ -328,6 +329,217 @@ unsigned ScanLoweringHelper::getScratchSizeInBytes() {
+  //   src = {register = [[1,0], [2,0]], lane = [[0,1], [0,2]]}
+  //   dst = {register = [[0,1], [2,0]], lane = [[1,0], [0,2]]}
+  //   cvt = dst, since src is the identity layout
+  //
+  // The map from destination -> source looks like:
+  //
+  //             dst_lane
+  // dst_reg       0      1      2      3
+  //  0          T0:0   T0:1   T2:0   T2:1
+  //  1          T1:0   T1:1   T3:0   T3:1
+  //  2          T0:2   T0:3   T2:2   T2:3
+  //  3          T1:2   T1:3   T3:2   T3:3
+  //
+  // Note for each destination register, two lanes want two different registers
+  // in the same source lane (T0:0 -> T0:0, T1:0 -> T0:1). This is impossible to
+  // represent with a warp shuffle, because the source lane (e.g. T0) can only
+  // supply one of its registers as the shuffle value.
+  //
+  // The goal of `P2` is to permute registers within a thread so that this does
+  // not happen. Specifically, pick `P2` such that bases in
+  // `(P2^-1 ∘ C).sublayout(kLane, {kLane, kRegister})` has non-zero lane
+  // components when the register components are non-zero.
+  //
+  // P2 can only change the register mapping within a thread. Constrain P2 as:
+  //
+  //   P2 = [ I P ]
+  //        [ P I ]
+  //
+  // Then `P2^-1 ∘ C` is:
+  //
+  //   [ I  0 ] [ C(r,r) C(r,l) ] = [ C(r,r)             C(r,l)           ]
+  //   [ P' I ] [ C(l,r) C(l,l) ]   [ P'*C(r,r)+C(l,r)   P'*C(r,l)+C(l,l) ]
+  //
+  // Where addition in GF(2) is xor.
+  //
+  // We can see that P' selects rows (i.e. bases) from the upper half (register)
+  // and combines them with the lower half (lane). Because the goal is for P' to
+  // select register bases `i` where C(r,l)[i] != 0, we know P'*C(r,r) = 0,
+  // since the corresponding C(r,r)[i] element in the same row will be zero.
+  //
+  // Note that solutions for P' do not always exist (no register permutation
+  // will decompose C to make the warp shuffle possible), and this happens when
+  // there aren't enough non-zero bases in C(r,l).
+  //
+  // Find the indices of the missing lane bases: rows in the lower half where
+  // the register component is non-zero but the lane component is zero.
+  SmallVector<int> missingLaneRows;
+  for (int i : llvm::seq(C.getInDimSizeLog2(kLane))) {
+    ArrayRef<int32_t> /*C(l,(r,l))[i]*/ lowerHalfRow = C.getBasis(kLane, i);
+    assert(lowerHalfRow.size() == 2);
+    if (/*C(l,r)*/ lowerHalfRow[0] != 0) {
+      assert(/*C(l,l)[i]*/ lowerHalfRow[1] == 0);
+      missingLaneRows.push_back(i);
+    } else if (lowerHalfRow[1] == 0) {
+      // If there is broadcasting along the lane, then C'(l,l) below won't be
+      // invertible. Intuitively, the dst tensor contains a subset of the src
+      // tensor's data, so recovering the src tensor through permutation alone
+      // is impossible. We would need an affine component (bfly shuffle).
+      return {};
+    }
+  }
+
+  // Find rows in the upperhalf that can be selected by P' to make the lane
+  // components in the lower half non-zero.
```

Comment:
I don't understand why it's called upper half and lower half. Isn't it left and right?

Reply:
I was referring to

```
  //   [ I  0 ] [ C(r,r) C(r,l) ] = [ C(r,r)             C(r,l)           ]
  //   [ P' I ] [ C(l,r) C(l,l) ]   [ P'*C(r,r)+C(l,r)   P'*C(r,l)+C(l,l) ]
```

Reply:
Oh, I get it. Never mind. It is the lower and upper from the perspective of `C`